straxen/plugins/nveto_recorder.py

```diff
@@ -0,0 +1,374 @@
+    each other then the specified resolving time.
+
+    Note:
+        1.) For the nVETO recorder we treat every fragment as a single
+            signal.
+        2.) The coincidence window in here is not self extending. Hence
+            we compute only the start times of a coincidence window.
+        3.) By default we store the last nfold - 1 hits, since there
+            might be an overlap to the next chunk. If we can not save
+            break the chunks.
+
+    Args:
+        rr (raw_records): raw_records
+        nfold (int): coincidence level.
+        resolving_time (int): Time window of the coincidence [ns].
+
+    Returns:
+        np.array: array containing the start times of the n-fold
+        coincidence intervals
+    """
+    # 1. estimate time difference between fragments:
+    start_times = rr['time']
+    mask = np.ones(len(start_times), dtype=np.bool_)
+    t_diff = diff(start_times)
+
+    # 2. Now we have to check if n-events are within resolving time:
+    #   -> use moving average with size n to accumulate time between n-pulses
+    #   -> check if accumulated time is below resolving time
+
+    # generate kernel:
+    kernel = np.zeros(nfold)
+    kernel[:(nfold - 1)] = 1  # weight last seen by t_diff must be zero since
+    # starting time point e.g. n=4: [0,1,1,1] --> [f1, f2, f3, f4, ..., fn]
+
+    t_cum = convolve1d(t_diff, kernel, mode='constant', origin=(nfold - 1) // 2)
+    t_cum = t_cum[:-(nfold - 1)]  # do not have to check for last < nfold hits
+    # since we will store these rr anyhow...
+    mask[:-(nfold - 1)] = t_cum <= resolving_time
+
+    return start_times[mask]
+
+
+@numba.njit(nogil=True, cache=True)
+def _merge_intervals(start_time, resolving_time):
+    """
+    Function which merges overlapping time intervals into a single one.
+
+    Args:
+        start_time (np.array): Start time of the different intervals
+        resolving_time (int): Coincidence window in ns
+
+    Returns:
+        np.array: merged unique time intervals.
+    """
+    # check for gaps larger than resolving_time:
+    # The gaps will indicate the starts of new intervals
+    gaps = diff(start_time) > resolving_time
+    interval_starts = np.arange(0, len(gaps), 1)
+    interval_starts = interval_starts[gaps]
+
+    # Creating output
+    # There is one more interval than gaps
+    intervals = np.zeros((np.sum(gaps) + 1, 2), dtype=np.int64)
+
+    # Looping over all intervals, except for the last one:
+    for ind, csi in enumerate(interval_starts[:-1]):
+        nsi = interval_starts[ind + 1]
+        intervals[ind + 1] = [start_time[csi], start_time[nsi - 1] + resolving_time]
+
+    # Now doing the first and last one:
+    intervals[0] = [start_time[0], start_time[interval_starts[0] - 1] + resolving_time]
+    intervals[-1] = [start_time[interval_starts[-1]], start_time[-1] + resolving_time]
+
+    return intervals
+
+
+@numba.njit(parallel=True, nogil=True)
+def diff(array):
```

Comment:
is np.diff noticeably slower?

Reply:
Depends on the array size: 
- 1000 elements: Numpy: 17.8 µs, Numba: 2.3 µs
- 100000 elements: Numpy 163 µs, Numba: 197 µs
Which means, that for our purposess PMTs x Dark Rate x Chunksize ~ 1800000 you are right. Numpy is even faster than my numba function. The main reason I wrote this function is, that I wanted to have a diff_function which preserves the array shape with a leading zero as a first entry. I will change it.

Reply:
Numba also supports np.diff directly inside jitted functions: https://numba.pydata.org/numba-doc/dev/developer/autogen_numpy_listing.html#numpy.diff.

Reply:
Done